### PR TITLE
A refactoring of the rollback code...

### LINKED ISSUE #25: redeploy scheme improvement
Current redeploy is very conservative and quite slow (redeploys one node at a time).  An alternative scheme might be to deploy an entirely new cluster (with `canary=start`), which would effectively create a blue/green deployment.  Then (on `canary=finish`), run the 'predelete' callback to sync-out the old cluster (then perhaps a new `canary=clean` to delete the old cluster).

### PR DESCRIPTION
**Pull request by dseeley**
...which also necessitated a refactor of some other variables.  Speed is considerably improved in all circumstances, and a couple of edge cases are fixed.  Breaks the rollback interface - predelete_role now receives a list of VMs `hosts_to_remove` to delete, rather than a single VM (was `host_to_redeploy`).  This has potential to massively increase redeploy speed, if clustering affinity is configured.

+ A new VM tag/label `lifecycle_state` is created describing the lifecycle state of the VM.  It is either `current`, `retiring` or `redeployfail`.
  + A cluster`s VMs will now always have the same epoch suffix (even when adding to the cluster)
  + The `-e clean` functionality has changed; you can now either clean hosts in every `lifecycle_state` (`-e clean=all`), or optionally just the VMs in one of the above states (`-e clean=retiring`).
  + Redeploy will assert if the topology has changed (number of VMs does not match reality)

+ A new global fact `cluster_hosts_state` is created that contains information on all running VMs with the derived cluster_name; i.e. the _state_ of the cluster.
  + Variables in `cluster_hosts_state` are used instead of constantly querying the infrastructure, esp during redeploy.

+ Alternate redeploy scheme: `_scheme_addallnew_rmdisk_rollback`.
  + A full mirror of the cluster is deployed.
  + If the process proceeds correctly:
    + `predeleterole` is called with a _list_ of the old VMs, in `hosts_to_remove`.
    + The old VMs are stopped.
  + If the process fails for any reason, the old VMs are reinstated, and the new VMs stopped (rollback)
  + To delete the old VMs, either set `-e canary_tidy_on_success=true`, or call redeploy.yml with `-e canary=tidy`

+ The existing `_scheme_addnewvm_rmdisk_rollback` scheme is refactored to use the new variables. It is functionally similar, but does not terminate the VMs on success.
  + For each node in the cluster:
    + Create a new VM
    + Run `predeleterole` on the previous node as a _list_ (for compatibility), in `hosts_to_remove`.
    + Shut down the previous node.
  + If the process fails for any reason, the old VMs are reinstated, and any new VMs that were built are stopped (rollback)
  + To delete the old VMs, either set `-e canary_tidy_on_success=true`, or call redeploy.yml with `-e canary=tidy`

Fixes #25

### CODE REVIEW

File: config/tasks/create_dns_a.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  register: dig_result
+  until: "dig_result.stdout == new_ip"
+  retries: 31
+  delay: 10
+  with_items: "{{ cluster_hosts_flat }}"
+  vars:
+    new_fqdn: "{{item.hostname}}.{{cluster_vars.dns_zone_external}}."
+    new_ip: "{{hostvars[item.hostname].ansible_host}}"
+  become: false
+  delegate_to: localhost
```
Comment (dipiazzg) [inline config/tasks/create_dns_a.yml]: `delegate_to: localhost` should be removed to make the play working even if it's running on a bastion

File: config/tasks/create_dns_a.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        target: "{%- if r__gcp_dns_managed_zone_info.resources.0.visibility == 'private' -%} {{ hostvars[item.hostname]['ansible_default_ipv4']['address'] }} {%- else -%} {{ hostvars[item.hostname]['ansible_host'] }} {%- endif -%}"
+        type: A
+        ttl: 60
+      become: false
+      delegate_to: localhost
+      run_once: true
+      with_items: "{{ cluster_hosts_flat }}"
+  when: cluster_vars.dns_server=="clouddns"
+
+
```
Comment (dipiazzg) [inline config/tasks/create_dns_a.yml]: looks like route53 takes a few seconds to create a DNS record. That means that `dig` a record which doesn't exist yet generate negative cache which makes the playbook failing.
The below fixes the issue:
```
- name: config/dns/a/dig | sleep for 10 seconds and continue with play
  wait_for:
    timeout: 10
  delegate_to: localhost
  run_once: true

File: readiness/tasks/config_dns_cname.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  delegate_to: localhost
+  run_once: true
+  when: cluster_vars.dns_server == "nsupdate"
+
+- name: create/dns/cname/route53 | create/update CNAME records in AWS (route53)
+  route53:
+    aws_access_key: "{{cluster_vars[buildenv].aws_access_key}}"
+    aws_secret_key: "{{cluster_vars[buildenv].aws_secret_key}}"
+    state: present
+    zone: "{{dns_tld_external}}"
```
Comment (dipiazzg) [inline readiness/tasks/config_dns_cname.yml]: should be:
`zone: "{{cluster_vars.dns_zone_external}}"`

File: clean/tasks/clean_dns.yml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      when: cluster_vars.dns_server == "nsupdate"
+
+    - name: clean/dns/route53 | Delete DNS entries from route53
+      block:
+        - name: clean/dns/route53 | Get A records
+          route53:
+            aws_access_key: "{{cluster_vars[buildenv].aws_access_key}}"
+            aws_secret_key: "{{cluster_vars[buildenv].aws_secret_key}}"
+            state: "get"
+            zone: "{{dns_tld_external}}"
```
[the remark this answers is not in the document]
Comment (dseeley-sky) [reply clean/tasks/clean_dns.yml]: I'm not so sure... `zone:` needs to be set to the root zone as defined in AWS or BIND or clouddns; this should be what `dns_tld_external` should be set to.  `dns_zone_external` is not this, it prepends a list of extra subdomains to `dns_tld_external`.

File: config/tasks/create_dns_a.yml (hunk 2 above)
Comment (dseeley-sky) [reply config/tasks/create_dns_a.yml]: Good explanation as to why that happens - will fix.

File: config/tasks/create_dns_a.yml (hunk 1 above)
Comment (dseeley-sky) [reply config/tasks/create_dns_a.yml]: Ok

Comment (dseeley-sky) [reply config/tasks/create_dns_a.yml]: In removing this, the command will run on one of the inventory hosts, which are inside the cloud (e.g.AWS).  DNS resolution inside AWS will always return the _VPC_ IP address; if you're using external IPs, they won't resolve.

**Comment by dseeley** [comment]
> Could `8.8.8.8` be a variable in `cluster_vars` to give end-users the choice of a public resolver that is not Google's one?

Added an option for this.

File: _dependencies/defaults/main.yml [hunk 5]
```diff
@@ -1,6 +1,4 @@
 ---
-# Ansible interpreter
-ansible_python_interpreter: "/usr/bin/python3"
```
Comment (eolix) [inline _dependencies/defaults/main.yml:3]: When running a redeploy, the absence of a defined interpreter resulted in an error: 
```
[REDACTED] \"module_stderr\": \"/bin/sh: 1: /usr/bin/python: not found\\n\", \"module_stdout\": \"\", \"msg\": \"The module failed to execute correctly, you probably need to set the interpreter.\\nSee stdout/stderr for the exact error\", \"rc\": 127}",
```

Any specific reason why this line was removed?

Comment (dseeley) [reply _dependencies/defaults/main.yml:3]: It should not be necessary to define the interpreter - Ansible should look it up automatically (https://docs.ansible.com/ansible/latest/reference_appendices/interpreter_discovery.html).  It was only there to _force_ Python3 for some earlier testing, and should not be there in general.  I've found that this error is a symptom of something else; did you manage to _deploy_ without the interpreter set?

**Pull request merged**